text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,485 @@
+
+We prefer the option of specifying behavior in the library using the normal trait, though it's a compatibility break for users of Rust who don't adopt the `core` crate (including compiler tests).
+
+## Implement for `Weak` and `NonNull`
+
+`Weak<T>` and `NonNull<T>` were not supported by the prior unstable arbitrary self types support, but they share the property that it may be desirable to implement method calls to `T` using them as self types. Unfortunately they also share the property that these types have many Rust methods using `self`, `&self` or `&mut self`. If we added to the set of Rust methods in future, we'd [shadow any such method calls](#method-shadowing). We can't implement `Receiver` for these types unless we come up with a policy that all subsequent additions to these types would instead be associated functions. That would make the future APIs for these types a confusing mismash of methods and associated functions, and the extra user complexity doesn't seem merited.
+
+## Not do it
+[not-do-it]: #not-do-it
+
+As always there is the option to not do this. But this feature already kind of half-exists (we are talking about `Box`, `Pin` etc.) and it makes a lot of sense to also take the last step and therefore enable non-libstd types to be used as self types.
+
+There is the option of using traits to fill a similar role, e.g.
+
+```rust
+trait ForeignLanguageRef {
+    type Pointee;
+    fn read(&self) -> *const Self::Pointee;
+    fn write(&mut self, value: *const Self::Pointee);
+}
+
+// --------------------------------------------------------
+
+struct ConcreteForeignLanguageRef<T>(T);
+
+impl<T> ForeignLanguageRef for ConcreteForeignLanguageRef<T> {
+    type Pointee = T;
+
+    fn read(&self) -> *const Self::Pointee {
+        todo!()
+    }
+
+    fn write(&mut self, _value: *const Self::Pointee) {
+        todo!()
+    }
+}
+
+// --------------------------------------------------------
+
+struct SomeForeignLanguageType;
+
+impl ConcreteForeignLanguageRef<SomeForeignLanguageType> {
+    fn m(&self) {
+        todo!()
+    }
+}
+
+trait Tr {
+    type RustType;
+
+    fn tm(self)
+    where
+        Self: ForeignLanguageRef<Pointee = Self::RustType>;
+}
+
+impl Tr for ConcreteForeignLanguageRef<SomeForeignLanguageType> {
+    type RustType = SomeForeignLanguageType;
+    fn tm(self) {}
+}
+
+fn main() {
+    let a = ConcreteForeignLanguageRef(SomeForeignLanguageType);
+    a.m();
+    a.tm();
+}
+```
+
+This successfully allows method calls to `m()` and even `tm()` without a reference to a `SomeForeignLanguageType` ever existing. However, due to the orphan rule, this forces every crate to have its own equivalent of `ConcreteForeignLanguageRef`. This workaround has been used by some interop tools, but use across multiple crates requires many generic parameters (`impl ForeignLanguageRef<Pointee=SomeForeignLanguageType>`).
+
+## Always use `unsafe` when interacting with other languages
```

Comment:
I'd be happy to drop it, because discussion of FFI unsafety sometimes leads to a fair bit of bikeshedding.

However I'm aiming to head off a kind of FAQ here: "Why do you need all this? Can't you just use regular references and use `unsafe` to ensure they're correct? Then we can all just use `Deref` and we don't need any of this `Receiver` stuff."

Superficially I think that's an appealing option. It only falls apart when you realize that humans also can't guarantee the semantics of C++ references match Rust, because they simply don't.

So, that's why I put it in. If you still think it does more harm than good, I'll zap it!